# 1 配置

验证失败，会抛出`ValidationException`

## 1.1 dependency

```
implementation("org.hibernate.validator:hibernate-validator:8.0.1.Final")
implementation("org.glassfish.expressly:expressly:5.0.0")
```

## 1.2 快速失败模式

默认会验证所有属性。改为`failFast`模式，遇到第一个违规的属性，立即抛出异常。

```kotlin
@Bean
fun validator(): Validator = Validation
    .byProvider(HibernateValidator::class.java)
    .configure()
    .failFast(true)
    .buildValidatorFactory().validator
```

## 1.3 Controller

`@Validated`, `@Valid`

```kotlin
@Validated
@Controller
class UserController {
    private val log = LoggerFactory.getLogger(this::class.java)

    @MessageMapping("anonymous.cash.transfer")
    suspend fun transfer(@Valid dto: TransferDto) {
        log.info("${dto.from}给${dto.to}转账${dto.amount}元")
    }
}
```

## 1.4 model

```kotlin
import jakarta.validation.constraints.Max
import kotlinx.serialization.Serializable

@Serializable
data class TransferDto(
    val from: String = "德刚",
    val to: String = "Andy",
    @field:Max(20, message = "转账金额 \${validatedValue} > {value}，账户余额不足")
    val amount: Int,
)
```

# 2 应用场景

| 场景      | 描述/示例                                                       |
| ------- | ----------------------------------------------------------- |
| 验证基本类型  | Int/String/Boolean等                                         |
| 验证class | data class Phone(val countryCode:String, val number:String) |
| 使用正则表达式 | 验证email、日期格式等                                               |
| 自定义注解   | 预定会议时间**段**                                                 |
